File: WinSet.vb
Public Class WinSet
    Friend SettingID As Byte
    Private dr As DataRow
    Private Sub WinSet_Shown() Handles Me.Shown
        Me.Text = WinTitleString()
        Call TransforConfig()
        Call InitControl()
        NudHour.Focus()
    End Sub
    Private Sub InitControl()
        If dr.Item("DtcConfiged") Then Call InitConfig() Else Call InitDefault()
    End Sub
    Private Sub InitDefault()
        RdbTime.Checked = True
        NudHour.Value = 0
        NudMinute.Value = 0
        NudSecond.Value = 0
        BtnON.Text = "&ON"
        BtnClear.Enabled = False
        BtnON.Tag = True.ToString
    End Sub
    Private Sub InitConfig()
        If dr.Item("DtcAsInteval") Then
            RdbInteval.Checked = True
            Dim ts As TimeSpan = dr.Item("DtcValueInteval")
            NudHour.Value = ts.Hours
            NudMinute.Value = ts.Minutes
            NudSecond.Value = ts.Seconds
        Else
            RdbTime.Checked = True
            Dim dt As DateTime = dr.Item("DtcValueTime")
            NudHour.Value = dt.Hour
            NudMinute.Value = dt.Minute
            NudSecond.Value = dt.Second
        End If
        BtnON.Text = If(dr.Item("DtcEnable"), "&OFF", "&ON")
        BtnClear.Enabled = True
    End Sub
    Private Sub TransforConfig()
        dr = WinMain.DtsMain.Tables("DttConfig").Select("DtcIDc=" & SettingID.ToString)(0)
    End Sub
    Private Function WinTitleString() As String
        Return "Alarm " & SettingID.ToString("X") & " - Multi Timer"
    End Function
    Private Sub NudHour_GotFocus(sender As NumericUpDown, e As EventArgs) Handles _
        NudHour.GotFocus, NudMinute.GotFocus, NudSecond.GotFocus
        sender.Select(0, sender.Value.ToString.Length)
    End Sub
    Private Sub WinSet_KeyDown(sender As Object, e As KeyEventArgs) Handles _
        Me.KeyDown, RdbTime.KeyDown, RdbInteval.KeyDown, NudHour.KeyDown, NudMinute.KeyDown, NudSecond.KeyDown
        Select Case e.KeyCode
            Case Keys.Escape
                Me.Close()
            Case Keys.Enter
                If dr.Item("DtcEnable") Then BtnSet.PerformClick() Else BtnON.PerformClick()
        End Select
    End Sub
    Private Sub WinSet_Closed() Handles Me.Closed
        Call ThemeMain.WinClockShow(WinMain.WinStyle)
    End Sub
    Private Sub BtnON_Click() Handles BtnON.Click
        Dim en As Boolean = Not dr.Item("DtcEnable")
        Call DataOpr.SetAlarm(SettingID, RdbInteval.Checked, NudHour.Value, NudMinute.Value, NudSecond.Value)
        Call DataOpr.AlarmOn(SettingID, en)
        Call WinMain.FlushAll()
        Me.Close()
    End Sub
    Private Sub BtnSet_Click() Handles BtnSet.Click
        Dim en As Boolean = dr.Item("DtcEnable")
        Call DataOpr.SetAlarm(SettingID, RdbInteval.Checked, NudHour.Value, NudMinute.Value, NudSecond.Value)
        Call DataOpr.AlarmOn(SettingID, en)
        Call WinMain.FlushAll()
        Me.Close()
    End Sub
    Private Sub BtnClear_Click() Handles BtnClear.Click
        Call DataOpr.ClearAlarm(SettingID)
        Call WinMain.FlushAll()
        Me.Close()
    End Sub
End Class

File: WinSet.cs
using System;
using System.Windows.Forms;

namespace Multi_Timer
{
    public partial class WinSet : Form
    {
        static public int intID;
        private Alarm almSet;
        private bool bolIniting;
        private int intH, intM, intS;
        public WinSet()
        {
            InitializeComponent();
        }
        private void WinSet_Shown(object sender, EventArgs e)
        {
            InitAlarm();
        }
        private void WinSet_Closed(object sender, FormClosedEventArgs e)
        {
            WinMain.bolSetting = false;
            this.Owner.Enabled = true;
            this.Owner.Activate();
            tmrSet.Enabled = false;            
        }
        private void InitAlarm()
        {
            bolIniting = true;
            this.Text = "Alarmer " + intID.ToString() + " - Multi Timer";
            almSet = WinMain.almMain[intID];
            switch (almSet.Configed)
            {
                case AlarmConfiged.Configed:
                    ChangeInit();
                    break;
                case AlarmConfiged.Unconfiged:
                default:
                    NewInit();
                    break;
            }            
            tmrSet.Enabled = true;
            FlushNote();
            nudH.Focus();            
            bolIniting = false;
        }
        private void ChangeInit()
        {
            almSet.Status = AlarmStatus.OFF;            
            switch (almSet.Type)
            {
                case AlarmType.Alarming:
                    rdbAlm.Checked = true;
                    nudH.Value = almSet.Alarming.Hour;
                    nudM.Value = almSet.Alarming.Minute;
                    nudS.Value = almSet.Alarming.Second;
                    break;
                case AlarmType.Distance:
                    rdbDst.Checked = true;
                    nudH.Value = almSet.Distance.Hours;
                    nudM.Value = almSet.Distance.Minutes;
                    nudS.Value = almSet.Distance.Seconds;
                    break;
                default:
                    NewInit();
                    break;
            }
            btnClr.Enabled = true;
        }
        private void NewInit()
        {
            almSet.Clear();
            rdbAlm.Checked = true;
            DateTime dt = DateTime.Now;
            almSet.Alarming = dt;
            nudH.Value = dt.Hour;
            nudM.Value = dt.Minute;
            nudS.Value = dt.Second;
            btnClr.Enabled = false;
        }
        public void FlushNote()
        {
            DateTime n = DateTime.Now;
            switch (almSet.Type)
            {
                case AlarmType.Alarming:
                    TimeSpan ts = almSet.Alarming - n;
                    if (ts.CompareTo(Common.tsZero) < 0)
                    {
                        ts += Common.tsOneDay;
                    }
                    intH = ts.Hours; intM = ts.Minutes; intS = ts.Seconds;
                    lblType.Text = "Alarming:";
                    break;
                case AlarmType.Distance:
                    DateTime dt = n + almSet.Distance;
                    intH = dt.Hour; intM = dt.Minute; intS = dt.Second;
                    lblType.Text = "Distance:";
                    break;
                default:
                    intH = 0; intM = 0; intS = 0;
                    lblType.Text = "Unknown:";
                    break;
            }
            lblTime.Text = intH.ToString().PadLeft(2, (char)48) + 
                ":" + intM.ToString().PadLeft(2, (char)48) + 
                ":" + intS.ToString().PadLeft(2, (char)48);
        }
        private void TmrSet_Tick(object sender, EventArgs e)
        {
            if (WinMain.bolSetting)
            {
                FlushNote();
            }
        }        
        private void BtnClr_Click(object sender, EventArgs e)
        {
            almSet.Clear();
            this.Close();
        }
        private void BntSet_Click(object sender, EventArgs e)
        {
            this.Close();
        }
        private void BtnON_Click(object sender, EventArgs e)
        {
            almSet.Status = AlarmStatus.OFF;
            almSet.Status = AlarmStatus.ON;           
            this.Close();
        }
        private void Nud_KeyPress(object sender, KeyPressEventArgs e)
        {
            switch (e.KeyChar )
            {
                case (char)13:
                    almSet.Status = AlarmStatus.OFF;
                    almSet.Status = AlarmStatus.ON;
                    this.Close();
                    break;
                case (char)27:
                    almSet.Status = AlarmStatus.OFF;
                    this.Close();
                    break;
            }
        }
        private void Nud_ValueChanged(object sender, EventArgs e)
        {
            switch (almSet.Type)
            {
                case AlarmType.Alarming:
                    LoadAlarming();
                    break;
                case AlarmType.Distance:
                    LoadDistance();
                    break;
                default:
                    break;
            }
            FlushNote();
        }
        private void RdbAlm_CheckedChanged(object sender, EventArgs e)
        {            
            if (((RadioButton)sender).Checked & !bolIniting)
            {
                LoadAlarming();
                InitAlarm();
            }
        }
        private void RdbDst_CheckedChanged(object sender, EventArgs e)
        {
            if (((RadioButton)sender).Checked & !bolIniting)
            {
                LoadDistance();
                InitAlarm();
            }            
        }
        private void Nud_GotFocus(object sender, EventArgs e)
        {
            NumericUpDown nud = (NumericUpDown)sender;
            nud.Select(0, 2);
        }
        private void LoadDistance()
        {
            almSet.Distance = new TimeSpan((int)nudH.Value, (int)nudM.Value, (int)nudS.Value);
        }
        private void LoadAlarming()
        {
            almSet.Alarming = Common.dtToday + (new TimeSpan((int)nudH.Value, (int)nudM.Value, (int)nudS.Value));
            if ((almSet.Alarming - DateTime.Now).CompareTo(Common.tsZero) < 0)
            {
                almSet.Alarming = almSet.Alarming.AddDays(1);
            }          
        }
    }
}
